tests/Mono.Android-Tests/jni/reuse-threads.c

```diff
@@ -200,3 +200,46 @@ rt_invoke_callback_on_new_thread (CB cb)
 	return 0;
 }
 
+static int
+_register_type_from_new_thread (const char *java_type_name, jobject class_loader)
+{
+	JNIEnv *env              = _get_env ("_register_type_from_new_thread");
+	jclass ClassLoader_class = (*env)->FindClass (env, "java/lang/ClassLoader");
+	jmethodID loadClass      = (*env)->GetMethodID (env, ClassLoader_class, "loadClass", "(Ljava/lang/String;)Ljava/lang/Class;");
+	jobject loaded_class     = (*env)->CallObjectMethod (env, class_loader, loadClass, java_type_name);
+
+	if ((*env)->ExceptionOccurred (env) != NULL) {
+		__android_log_print (ANDROID_LOG_INFO, "XA/RuntimeTest", "FAILURE: class '%s' cannot be loaded, Java exception thrown!", java_type_name);
+		return -1;
+	}
+
+	jmethodID Object_ctor    = (*env)->GetMethodID (env, loaded_class, "<init>", "()V");
+	jobject   instance       = (*env)->NewObject (env, loaded_class, Object_ctor);
+
+	if ((*env)->ExceptionOccurred (env) != NULL || instance == NULL) {
+		__android_log_print (ANDROID_LOG_INFO, "XA/RuntimeTest", "FAILURE: instance of class '%s' wasn't created!", java_type_name);
+		return -1;
+	}
+
+	return 0;
+}
+
+JNIEXPORT int JNICALL
+rt_register_type_on_new_thread (const char *java_type_name)
+{
+	JNIEnv *env = _get_env ("rt_register_type_on_new_thread");
+	pthread_t t;
+
+	/* 1: Create a thread... */
+	int r = pthread_create (&t, NULL, _register_type_from_new_thread, java_type_name);
+
+	if (r) {
+		__android_log_print (ANDROID_LOG_INFO, "XA/RuntimeTest", "RegisterOnNewThread: pthread_create() failed! %i: %s", r, strerror (r));
+		return -1;
+	}
+
+	void *tr;
+	pthread_join (t, &tr);
+
+	return 0;
```

Comment:
…though the `return` statements within `_register_type_from_new_thread()` are kinda moot, as we drop the return value that `pthread_join()` provides!

Which means that the unit test will always succeed, even if `ClassLoader.loadClass()` fails!  (There just might be some "random unrelated crash elsewhere" as the pending Java exception is actually thrown…